testing/debug/debug.c

```diff
@@ -0,0 +1,282 @@
+ * Name: debug_callback
+ ****************************************************************************/
+
+static void debug_callback(int type, FAR void *addr, size_t size,
+                           FAR void *arg)
+{
+  *(bool *)arg = true;
+  gdb_debugpoint_remove(type, addr, size);
+}
+
+/****************************************************************************
+ * Name: test_function
+ ****************************************************************************/
+
+static int test_function(void (*func)(void))
+{
+  int ret = 0;
+  bool triggered;
+  printf("==== Calling test_function ====\n");
+
+  printf("Add breakpoint at %p\n", func);
+  ret = gdb_debugpoint_add(GDB_STOPREASON_BREAKPOINT, func, 0,
+                          debug_callback, &triggered);
+
+  if (ret != OK)
+    {
+      printf("ERROR: Registering breakpoint at %p failed with %d\n",
+             func, ret);
+      return ret;
+    }
+
+  /* Trigger the breakpoint */
+
+  triggered = false;
+  func();
+
+  if (triggered)
+    {
+      printf("Pass: Breakpoint at %p triggered\n", func);
+    }
+  else
+    {
+      gdb_debugpoint_remove(GDB_STOPREASON_BREAKPOINT, func, 0);
+      printf("ERROR: Breakpoint %p not triggered\n", func);
+      ret = -1;
+    }
+
+  return ret;
+}
+
+/****************************************************************************
+ * Name: test_dataread
+ ****************************************************************************/
+
+static int test_data(uint8_t *addr, size_t size, bool r, bool w)
+{
+  int ret = 0;
+  int tmp;
+  int test_type;
+  bool triggered;
+
+  printf("==== Calling test_data ====\n");
+
+  if (r && w)
+    {
+      test_type = GDB_STOPREASON_WATCHPOINT_RW;
+      printf("Add read/write watchpoint at %p\n", addr);
+    }
+  else if (r)
+    {
+      test_type = GDB_STOPREASON_WATCHPOINT_RO;
+      printf("Add read watchpoint at %p\n", addr);
+    }
+  else if (w)
+    {
+      test_type = GDB_STOPREASON_WATCHPOINT_WO;
+      printf("Add write watchpoint at %p\n", addr);
+    }
+  else
+    {
+      printf("ERROR: Invalid test type\n");
+      return -1;
+    }
+
+  ret = gdb_debugpoint_add(test_type, addr, size,
+                          debug_callback, &triggered);
+
+  if (ret != OK)
+    {
+      printf("ERROR: Registering read watchpoint at %p failed with %d\n",
+             addr, ret);
+      return ret;
+    }
+
+  if (r & w)
+    {
+      /* Trigger the watchpoint by reading the address */
+
+      triggered = false;
+      if (addr[0] == 0x55)
+        {
+          /* Do something to avoid compiler removing the read */
+
+          tmp = 0xaa;
+        }
+      else
+        {
+          tmp = 0x55;
+        }
+
+      if (triggered)
+        {
+          printf("Pass: Read watchpoint at %p triggered\n", addr);
+        }
+      else
+        {
+          gdb_debugpoint_remove(test_type, addr, size);
+          printf("ERROR: Read watchpoint at %p not triggered\n", addr);
+          ret = -1;
```

Comment:
should we remove ALL `ret = -1` and `ret = 0`

Reply:
OK, removed